app/assets/src/components/views/PipelineViz.jsx

```diff
@@ -0,0 +1,379 @@
+          step.class = step.class.substring(12);
+        }
+        if (step.class.substring(0, 3) == "Run") {
+          step.class = step.class.substring(3);
+        }
+        if (step.class.substring(0, 8) == "Generate") {
+          step.class = step.class.substring(8);
+        }
+      });
+    });
+  }
+
+  generateNodeData(index, edgeData) {
+    const { stageResults } = this.props;
+    const stageData = stageResults.stages[this.stageNames[index]];
+    const stepData = stageData.steps;
+
+    const nodeData = [];
+    stepData.forEach((step, i) => {
+      nodeData.push({ id: i, label: step.class });
+    });
+    nodeData.push({ id: START_NODE_ID, group: "startEndNodes" });
+    nodeData.push({ id: END_NODE_ID, group: "startEndNodes" });
+
+    this.addHierarchicalLevelsToNodes(nodeData, edgeData);
+    return nodeData;
+  }
+
+  addHierarchicalLevelsToNodes(nodeData, edgeData) {
+    // This method assumes that the steps are topologically sorted already,
+    // in that each node's parent all appear in the array before it.
+    const nodeToCurrentLevel = {};
+    nodeData.forEach(node => {
+      nodeToCurrentLevel[node.id] = 1;
+    });
+
+    edgeData.forEach(edge => {
+      const newLevel = nodeToCurrentLevel[edge.from] + 1;
+      if (newLevel > nodeToCurrentLevel[edge.to]) {
+        nodeToCurrentLevel[edge.to] = newLevel;
+        if (edge.to != END_NODE_ID) {
+          nodeToCurrentLevel[END_NODE_ID] = Math.max(
+            nodeToCurrentLevel[END_NODE_ID],
+            newLevel + 1
+          );
+        }
+      }
+    });
+
+    nodeData.forEach(node => {
+      node.level = nodeToCurrentLevel[node.id];
+    });
+  }
+
+  generateIntraEdgeData(index) {
+    const { stageResults } = this.props;
+    const stageData = stageResults.stages[this.stageNames[index]];
+    const stepData = stageData.steps;
+
+    const outTargetToStepId = {};
+    stepData.forEach((step, i) => {
+      if (!(step.out in outTargetToStepId)) {
+        // Populate outFileToStepId for intra-stage edges
+        outTargetToStepId[step.out] = i;
+      }
+    });
+
+    const intraEdgeData = [];
+    stepData.forEach((step, i) => {
+      step.in.forEach(inTarget => {
+        if (inTarget in outTargetToStepId) {
+          const fromId = outTargetToStepId[inTarget];
+          intraEdgeData.push({ from: fromId, to: i });
+        } else {
+          // Connect beginning steps to input node
+          intraEdgeData.push({ from: START_NODE_ID, to: i });
+        }
+      });
+    });
+    return intraEdgeData;
+  }
+
+  generateInterEdgeData(index) {
+    const { stageResults, backgroundColor } = this.props;
+    const stageData = stageResults.stages[this.stageNames[index]];
+    const stepData = stageData.steps;
+
+    const interEdgeData = [];
+    if (index == this.stageNames.length - 1) {
+      // For final stage, create hidden edges to final node for vertical centering of nodes.
+      stepData.forEach((_, i) => {
```

Comment:
it looks like this is mutating props as well - can have unintended consequences.

Reply:
Should be resolved now — most (if not all) references to props have been changed!